### PR TITLE
#21282 adding the metadata interceptor

### LINKED ISSUE #21282: Identify which node a response comes from
## Is your feature request related to a problem? Please describe.

There are many types of support issues where identifying which node responded to a request is critical to resolving the issue (to identify a node which isn't responding correctly, identify out-of-sync nodes, etc.).  Currently there is no standard way to do this; it's been done by allowing access to each node in the cluster via separate URLs, modifying load balancer configuration, and other methods, but none of these methods works well, and none of them work universally for all systems.

This all means that each time this information is needed, a custom solution has to be implemented (and then usually rolled back after the problem has been solved).

## Describe the solution you'd like

We need a way for dotCMS to identify which node in a cluster each specific response is sent from.  This is what I suggest, but there are multiple ways this could be implemented:

* Add a response header which identifies which node the response was sent from.
* Enable or disable the response header to all responses via a configuration property (for example, `RESPONSE_HEADER_ADD_NODE_ID=true`).
* Allow enabling the response header in an individual request by adding a URL parameter (for example, `?nodeid=true`).

## Describe alternatives you've considered

dotCMS Cloud Engineering has considered implementing something of this nature on the load balancer for all dotCMS Cloud customers.  However, this isn't a workable option within our current infrastructure (ALB), and there are undoubtedly on-premise customers with similar limitations.

Other methods, such as bypassing the LB to access nodes directly, present performance and reliability issues, so they can't be implemented for any customers on more than a temporary (custom) basis.

## Additional context

The identifier used in the header doesn't matter, as long as:
* It's unique to each node.
* It can be easily used to identify the node within the infrastructure.

Ideally, the node identifier should change if the server or container running on the node is changed (especially with k8s, where multiple containers may be started from the same image).
* But if necessary, a unique ID (for the server/image) and a timestamp (for when the server/container was started) could potentially be both used to ensure identification of unique containers.

The Site from the license screen seems to fit most of these needs, so that might be an appropriate identifier to use.  If the Site doesn't change when k8s shuts down and restarts a container, then it might make sense to send a timestamp (or maybe Server ID) in addition to the Site.

### PR DESCRIPTION
**Pull request by jdotcms**
Adding a new web interceptor that basically may add meta info to the reponse

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @author jsanca
+ */
+public class MetaWebInterceptor implements WebInterceptor {
+
+    private static final String X_DOT_SERVER_HEADER = "x-dot-server";
+    private static final String FRIENDLY_NAME = "friendlyName";
+    private static final String UNKNOWN = "unknown";
+    private final Lazy<Boolean> responseHeaderAddNodeId = Lazy.of(()-> Config.getBooleanProperty("RESPONSE_HEADER_ADD_NODE_ID", true));
+    private final Lazy<String>  serverId                = Lazy.of(()-> StringUtils.shortify(APILocator.getServerAPI().readServerId(), 10));
+    private final Lazy<Map<String, Serializable>> cacheInfoMap = Lazy.of(() -> Try.of(() -> ClusterUtilProxy.getNodeInfo()).getOrElse(Collections.emptyMap()));
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptor.java:33]: I would just cache the `FriendlyName` as a String and not the whole NodeMap, which can get big.  Also, the inclusion of the `FriendlyName` should be configurable.

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptor.java:33]: ok

File: dotCMS/src/main/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Interceptor that adds meta info to the dotcms response
+ * @author jsanca
+ */
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptor.java:22]: What meta info it add?, I think this doc is incomplete

File: dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    @Test
+    public void check_header_x_dot_server()  {
+
+        final MetaWebInterceptor metaWebInterceptor = new MetaWebInterceptor();
+        final MockHeaderResponse mockHeaderResponse = new MockHeaderResponse(response);
+        metaWebInterceptor.intercept(request, mockHeaderResponse);
+
+        final String header = mockHeaderResponse.getHeader(MetaWebInterceptor.X_DOT_SERVER_HEADER);
+        Assert.assertNotNull(header);
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java:46]: shouldn't we check the value expected too?

**Review by freddyucv — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    public void check_header_x_dot_server_config_disable()  {
+
+        Config.setProperty(ResponseMetaDataWebInterceptor.RESPONSE_HEADER_ADD_NODE_ID, false);
+        final ResponseMetaDataWebInterceptor metaWebInterceptor = new ResponseMetaDataWebInterceptor();
+        final MockHeaderResponse mockHeaderResponse = new MockHeaderResponse(response);
+        metaWebInterceptor.intercept(request, mockHeaderResponse);
+
+        final String header = mockHeaderResponse.getHeader(ResponseMetaDataWebInterceptor.X_DOT_SERVER_HEADER);
+        Assert.assertNull(header);
+        Config.setProperty(ResponseMetaDataWebInterceptor.RESPONSE_HEADER_ADD_NODE_ID, true);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java:78]: if something fails before this line, the property would remain as `false`. Shouldn't this line be in a finally statement?

File: dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        Config.setProperty(ResponseMetaDataWebInterceptor.RESPONSE_HEADER_ADD_NODE_ID_INCLUDE_NODE_NAME, false);
+        final ResponseMetaDataWebInterceptor metaWebInterceptor = new ResponseMetaDataWebInterceptor();
+        final MockHeaderResponse mockHeaderResponse = new MockHeaderResponse(response);
+        metaWebInterceptor.intercept(request, mockHeaderResponse);
+
+        final String header = mockHeaderResponse.getHeader(ResponseMetaDataWebInterceptor.X_DOT_SERVER_HEADER);
+        Assert.assertNotNull(header);
+        Assert.assertTrue(header.startsWith("unknown|"));
+        Config.setProperty(ResponseMetaDataWebInterceptor.RESPONSE_HEADER_ADD_NODE_ID_INCLUDE_NODE_NAME, true);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java:99]: same here

File: dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.IOException;
+import java.util.Collections;
+
+import static org.mockito.Mockito.mock;
+
+/**
+ * Unit test for {@link ResponseMetaDataWebInterceptor}
+ * @author jsanca
+ */
+public class MetaWebInterceptorTest {
```
Comment (erickgonzalez) [inline dotCMS/src/integration-test/java/com/dotcms/filters/interceptor/meta/MetaWebInterceptorTest.java:26]: This test must be added to the MainSuite so it runs

**Review by erickgonzalez — CHANGES_REQUESTED** [review]

**Comment by erickgonzalez** [comment]
One of the tests created is failing https://raw.githack.com/dotCMS/test-results/issue-21282-response-node-header/projects/core/current/integration/postgres/reports/html/classes/com.dotcms.filters.interceptor.meta.MetaWebInterceptorTest.html#check_header_x_dot_server

**Review by erickgonzalez — APPROVED** [review]

**Pull request merged**